src/toxcore/dht/server.rs

```diff
@@ -123,35 +137,42 @@ impl Server {
         Box::new(pings_stream.for_each(|()| Ok(())))
     }
     /// send PingRequests to peers at every 60 seconds
-    pub fn send_ping_req(&mut self, peer: PackedNode) -> IoFuture<()> {
+    pub fn send_ping_req(&self, peer: PackedNode) -> IoFuture<()> {
         let mut client = self.create_client(&peer.saddr, peer.pk);
         let result = client.send_ping_request();
-        self.peers_cache.insert(peer.pk, client);
+        let mut state = self.state.write().expect("Failed to obtain ServerState lock");
+        state.peers_cache.insert(peer.pk, client);
         result
     }
     /// send NodesRequest to random peer at every 20 seconds
-    pub fn send_nodes_req(&mut self, friend_pk: PublicKey) -> IoFuture<()> {
-        if let Some(peer) = self.kbucket.get_random_node() {
+    pub fn send_nodes_req(&self, friend_pk: PublicKey) -> IoFuture<()> {
+        let mut state = self.state.write().expect("Failed to obtain ServerState lock");
+        self.send_nodes_req_resource(friend_pk, &mut state)
+    }
+    fn send_nodes_req_resource(&self, friend_pk: PublicKey, state: &mut ServerState) -> IoFuture<()> {
```

Comment:
You don't need to create `send_nodes_req_resource` if you don't use it anywhere else. It is ok to get lock in `send_nodes_req` and put the code into the method right after lock.